src/random_ray/linear_source_domain.cpp

```diff
@@ -157,41 +161,74 @@ int64_t LinearSourceDomain::add_source_to_scalar_flux()
-    if (was_cell_hit) {
+    if (volume_iteration > 0.0) {
       n_hits++;
     }
 
+    // Check if an external source is present in this source region
+    bool external_source_present =
+      external_source_present_.size() && external_source_present_[sr];
+
+    // The volume treatment depends on the volume estimator type
+    // and whether or not an external source is present in the cell.
+    double volume;
+    switch (volume_estimator_) {
+    case RandomRayVolumeEstimator::NAIVE:
+      volume = volume_iteration;
+      break;
+    case RandomRayVolumeEstimator::SIMULATION_AVERAGED:
+      volume = volume_simulation_avg;
+      break;
+    case RandomRayVolumeEstimator::HYBRID:
+      if (external_source_present) {
+        volume = volume_iteration;
+      } else {
+        volume = volume_simulation_avg;
+      }
+      break;
+    default:
+      fatal_error("Invalid volume estimator type");
+    }
+
     for (int g = 0; g < negroups_; g++) {
       int64_t idx = (sr * negroups_) + g;
       // There are three scenarios we need to consider:
-      if (was_cell_hit) {
+      if (volume_iteration > 0.0) {
         // 1. If the FSR was hit this iteration, then the new flux is equal to
         // the flat source from the previous iteration plus the contributions
         // from rays passing through the source region (computed during the
         // transport sweep)
         scalar_flux_new_[idx] /= volume;
         scalar_flux_new_[idx] += source_[idx];
         flux_moments_new_[idx] *= (1.0 / volume);
-      } else if (volume > 0.0) {
+      } else if (volume_simulation_avg > 0.0) {
         // 2. If the FSR was not hit this iteration, but has been hit some
-        // previous iteration, then we simply set the new scalar flux to be
-        // equal to the contribution from the flat source alone.
-        scalar_flux_new_[idx] = source_[idx];
+        // previous iteration, then we need to make a choice about what
+        // to do. Naively we will usually want to set the flux to be equal
+        // to the reduced source. However, in fixed source problems where
+        // there is a strong external source present in the cell, and where
+        // the cell has a very low cross section, this approximation will
+        // cause a huge upward bias in the flux estimate of the cell (in these
+        // conditions, the flux estimate can be orders of magnitude too large).
+        // Thus, to avoid this bias, if any external source is present
+        // in the cell we will use the previous iteration's flux estimate. This
+        // injects a small degree of correlation into the simulation, but this
+        // is going to be trivial when the miss rate is a few percent or less.
+        if (external_source_present) {
+          scalar_flux_new_[idx] = scalar_flux_old_[idx];
+          flux_moments_new_[idx] = flux_moments_old_[idx];
+        } else {
+          scalar_flux_new_[idx] = source_[idx];
+        }
```

Comment:
The `LinearSourceDomain::add_source_to_scalar_flux()` function is more or less identical to the `FlatSourceDomain::add_source_to_scalar_flux()` function. This seems to be an area ripe for inheritance, or at least some shared code. Looks liek the main difference is the storing and updating of the flux moments in the `LinearSourceDomain` version. I could be difficult to reuse the code within loop though without adding a bunch of checks.

Reply:
This is a great point -- the logic is very important here (as is the explanation of the logic), so would be best not to do it all twice. I'll work on finding a cleaner way of doing this.

Reply:
I've updated the branch so that there the outer `add_source_to_scalar_flux` function is no longer virtual. Instead of having `LinearSourceDomain` re-implement the entire function, there are now three very small methods (two of which are virtual):

```C++
  virtual void set_flux_to_flux_plus_source(int64_t idx, double volume, int material, int g);
  void set_flux_to_source(int64_t idx);
  virtual void set_flux_to_old_flux(int64_t idx);
```

 that are just a few lines each and handle the flux manipulation needed for the various cases. This means the general logic is no longer repeated. Note that `set_flux_to_source()` is not virtual as if a region is missed in linear source mode, we are currently just keeping the flat contribution but discarding the source moments. Early work so far makes it seem like keeping the source moment contribution is not helpful (and is not super cheap as we have to invert/solve with the moments matrix to convert the source moments to flux moments), but this policy may be revisited in the future.